src/hooks/useImmediateState.ts

```diff
@@ -0,0 +1,33 @@
+import { useRef, useEffect, useState, useCallback } from 'react';
+
+type SetStateAction<S> = S | ((prevState: S) => S);
+type SetState<S> = (action: SetStateAction<S>) => void;
+
+/** This is a drop-in replacement for useState, limited to object-based state. During initial mount it will mutably update the state, instead of scheduling a React update using setState */
+export const useImmediateState = <S extends {}>(init: S): [S, SetState<S>] => {
+  const isMounted = useRef(false);
+  const initialState = useRef<S>({ ...init });
+  const [state, setState] = useState<S>(initialState.current);
+
+  // This wraps setState and updates the state mutably on initial mount
+  // It also prevents setting the state when the component is unmounted
+  const updateState: SetState<S> = useCallback((action: SetStateAction<S>) => {
+    if (isMounted.current) {
+      setState(action);
+    } else if (typeof action === 'function') {
```

Comment:
Can we return from these conditions rather than having to parse multiple conditions.

```tsx
if (isMounted.current) {
  return setState(action);
}

const actionValue = typeof action === "function" ? action(initialState.current) : action;
Object.assign(initialState.current, actionValue);
```

Reply:
Doesn't really make a difference though so not to be too bikesheddy, but since we just follow some loose eslint rules,
you can always change it to what you prefer when you happen to refactor or rewrite this part of the codebase 😅